File: abstr1/update_site.py
from urllib.parse import urljoin

from abstr1.hosts import get_hosts
from abstr2.hosts import add_to_unknown_hosts, load_hosts, create_host_specific_index, start_scan_index, \
    stop_scan_index, update_scan_index_progress
from abstr2.index import add_to_index, is_etag_in_index, get_urls
from abstr2.logs import log_processing
from abstr2.robots import do_robot_delay, load_robots_parser
from abstr2.site_processing import etag_head
from abstr2.urls import needs_be_indexed
from snail_pipes.URLInput import process
from snail_pipes.url_filters import URLFilter
from abstr2.context import ContextMap


def crawl(host, contextmap: ContextMap):
    cached = 0
    processed = 0
    visited = 0
    urls_to_process = [host]
    start_scan_index(contextmap, host)
    all_urls = get_urls(contextmap)
    print(f"processing {host}")
    for url in all_urls:
        contextmap.current_url = url
        update_scan_index_progress(contextmap, host, visited * 100 / len(all_urls))
        visited += 1
        if not needs_be_indexed(contextmap):
            add_to_unknown_hosts(contextmap)
            try:
                urls_to_process.remove(url)
            except:
                continue
            continue
        tag = etag_head(url)
        in_database = is_etag_in_index(url, tag, contextmap)
        if in_database:
            try:
                urls_to_process.remove(url)
            except:
                continue
            cached += 1
            continue
        process_result = process(url)
        add_to_index(contextmap, process_result)
        for anchor in process_result.get_anchor_urls(url):
            if anchor not in urls_to_process:
                urls_to_process.append(anchor)
        processed += 1
        do_robot_delay(contextmap)
        log_processing(visited, urls_to_process, cached, processed)
    stop_scan_index(contextmap, host)
    return visited


def create_index_for_hostname(hostname, contextmap: ContextMap, starturl=None):
    contextmap.current_host = hostname
    contextmap = create_host_specific_index(contextmap)
    contextmap = load_hosts(contextmap)
    contextmap = load_robots_parser(contextmap)
    if starturl == None:
        starturl = f"https://{hostname}"
    contextmap.url_filter = URLFilter(f"https://{hostname}", [])
    print(crawl(starturl, contextmap))

for host in get_hosts():
    contextmap = ContextMap()
    create_index_for_hostname(host.replace(".db", ""), contextmap)

File: src/main/java/search/UpdateSite.java
package search;

import java.util.ArrayList;
import java.util.List;

/**
 * Port of abstr1.update_site.
 * Updates existing site index: iterates over URLs already in the index, re-checks etag,
 * re-processes changed pages and discovers new anchors. Contrast with ProcessNewSite
 * which does BFS from a single start URL.
 */
public final class UpdateSite {

    private static final Hosts HOSTS = new Hosts();
    private static final Index INDEX = new Index();
    private static final Urls URLS = new Urls();

    private UpdateSite() {
    }

    public static int crawl(String startUrl, ContextMap contextmap) {
        int cached = 0;
        int processed = 0;
        int visited = 0;
        List<String> urlsToProcess = new ArrayList<>();
        urlsToProcess.add(startUrl);

        contextmap.hostDAO.startScanIndex(startUrl);
        List<String> allUrls = INDEX.getUrls(contextmap);

        System.out.println("processing " + startUrl);

        for (String url : allUrls) {
            contextmap.currentUrl = url;
            if (!allUrls.isEmpty()) {
                contextmap.hostDAO.updateScanIndexProgress(startUrl, visited * 100.0 / allUrls.size());
            }
            visited++;

            if (!URLS.needsBeIndexed(contextmap)) {
                HOSTS.addToUnknownHosts(contextmap);
                urlsToProcess.remove(url);
                continue;
            }

            String tag = SiteProcessing.etagHead(url);
            if (INDEX.isEtagInIndex(url, tag, contextmap)) {
                urlsToProcess.remove(url);
                cached++;
                continue;
            }

            ProcessResult processResult = URLInput.process(url);
            INDEX.addToIndex(contextmap, processResult);
            for (String anchor : processResult.getAnchorUrls(url)) {
                if (!urlsToProcess.contains(anchor)) {
                    urlsToProcess.add(anchor);
                }
            }
            processed++;
            Robots.doRobotDelay(contextmap);
            Logs.logProcessing(visited, urlsToProcess, cached, processed);
        }

        contextmap.hostDAO.stopScanIndex(startUrl);
        return visited;
    }

    public static void updateIndexForHostname(String hostname, ContextMap contextmap, String starturl) {
        contextmap.currentHost = hostname;
        contextmap = HOSTS.createHostSpecificIndex(contextmap);
        contextmap = HOSTS.loadHosts(contextmap);
        contextmap = Robots.loadRobotsParser(contextmap);
        if (starturl == null) {
            starturl = "https://" + hostname;
        }
        contextmap.urlFilter = new URLFilter("https://" + hostname, List.of());
        System.out.println(crawl(starturl, contextmap));
    }

    public static void main(String[] args) {
        String dataDir = "data";
        boolean useCsv = true;
        Hosts hosts = new Hosts(dataDir);
        List<String> hostnames = hosts.getHosts(useCsv);

        for (String host : hostnames) {
            ContextMap contextmap = new ContextMap();
            contextmap.useCsv = useCsv;
            updateIndexForHostname(host, contextmap, null);
        }
    }
}
